nixos/modules/tasks/auto-upgrade.nix

```
@@ -174,6 +252,233 @@ in
+
+              # Internal options and configuration, to make this work
+              options.desyncUpgradeScript = lib.mkOption {
+                # Derivation called to upgrade the desync files.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                visible = false;
+                type = lib.types.package;
+              };
+
+              options.upgradeFailedScript = lib.mkOption {
+                # Derivation called when the latest upgrade failed.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                visible = false;
+                type = lib.types.package;
+              };
+
+              config =
+                let
+                  desync =
+                    if builtins.pathExists /var/lib/nixos/desyncs/${name}.nix then
+                      import /var/lib/nixos/desyncs/${name}.nix
+                    else
+                      {
+                        # If we have no desync saved, it means that any configuration that builds
+                        # would necessarily be with the default path.
+                        # This can happen eg. on the `nixos-rebuild` that adds the desync configuration
+                        # to a nixos config, that would most likely not happen during an auto-upgrade.
+                        desyncAge = 0;
+                        path = config.defaultPath;
+                      };
+                in
+                {
+                  desyncAge = desync.desyncAge;
+                  path = desync.path;
+                  pkgs = config.pkgsFor desync.path;
+
+                  desyncUpgradeScript = pkgs.runCommand "nixos-desync-${name}" { } ''
+                    #!/bin/sh
+                    set -euo pipefail
+
+                    # Force dependencies on `requireBuilds`
+                    # ${lib.concatStringsSep " " (config.requireBuilds (config.pkgsFor (config.defaultPath)))}
+
+                    # Write the result in `$out` if this passed
+                    cat - > "$out" <<EOF
+                    {
+                      desyncAge = 0;
+                      path = builtins.toPath ${
+                        # Convert the now-validated path into an absolute path in the store, then into a string
+                        lib.strings.escapeNixString (builtins.path { path = config.defaultPath; })
+                      };
+                    }
+                    EOF
+                  '';
+
+                  upgradeFailedScript = pkgs.runCommand "nixos-upgrade-failed-${name}" { } ''
+                    #!/bin/sh
+                    set -euo pipefail
+
+                    ${lib.optionalString (config.desyncAge >= config.maxDesyncAge) ''
+                      echo "Did not get any successful build for ${name} in the past ${config.desyncAge} rebuilds!" >&2
```

```suggestion
                      echo "Did not get any successful build for ${name} in the past ${toString config.desyncAge} rebuilds!" >&2
```

`desyncAge` is an int, so it can't be interpolated into a string without first using `builtins.toString`.

---

Good catch! You're right that I never actually hit this case in my testing… someday nix will be a compiled language :smiling_face_with_tear: